Don't accept user names with extra spaces

040medien opened the pull request:
<!--
Thank you for your interest in contributing to OctoPrint, it's
highly appreciated!

Please make sure you have read the "guidelines for contributing" as
linked just above this form, there's a section on Pull Requests in there
as well which contains important information.

As a summary, please make sure you have ticked all points on this
checklist:
-->

  * [X] Your changes are not possible to do through a plugin and relevant
    to a large audience (ideally all users of OctoPrint)
  * [X] If your changes are large or otherwise disruptive: You have
    made sure your changes don't interfere with current development by
    talking it through with the maintainers, e.g. through a
    Brainstorming ticket
  * [X] Your PR targets OctoPrint's `devel` branch if it's a completely
    new feature, or `maintenance` if it's a bug fix or improvement of
    existing functionality for the current stable version (no PRs
    against `master` or anything else please)
  * [X] Your PR was opened from a custom branch on your repository
    (no PRs from your version of `master`, `maintenance`, or `devel`
    please), e.g. `dev/my_new_feature` or `fix/my_bugfix`
  * [X] Your PR only contains relevant changes: no unrelated files,
    no dead code, ideally only one commit - rebase and squash your PR
    if necessary!
  * [X] Your changes follow the existing coding style
  * [X] If your changes include style sheets: You have modified the
    `.less` source files, not the `.css` files (those are generated
    with `lessc`)
  * [X] You have tested your changes (please state how!) - ideally you
    have added unit tests
  * [X] You have run the existing unit tests against your changes and
    nothing broke
  * [x] You have added yourself to the `AUTHORS.md` file :)

<!--
Describe your PR further using the template provided below. The more
details the better!
-->

#### What does this PR do and why is it necessary?
Users who accidentally add a space while entering or pasting a user name during initial setup might not notice it and could be prevented from logging in.
In the setup wizard,  don't allow creating a user with a trailing or leading space in the user name. There are no error messages added, just the "Create Account" button becomes inactive. I.e. same behavior as empty user name.

#### How was it tested? How can it be tested by the reviewer?
Tested locally by starting with a fresh .octoprint folder

#### Any background context you want to provide?

#### What are the relevant tickets if any?
https://github.com/OctoPrint/OctoPrint/issues/4621


#### Screenshots (if appropriate)

#### Further notes

cp2004 commented:
Do you think it is worth changing the API to disallow creating a username with extra whitespace as well?

foosel commented:
See my comment: https://github.com/OctoPrint/OctoPrint/issues/4621#issuecomment-1239035542:
> Agreed, and the input fields for creation in wizard and settings should reflect that accordingly by switching to error state if there's leading or trailing whitespace. Or whitespace in the first place.

040medien commented:
Thank you for the comments, I:
- introduced API validation
- introduced frontend error states and messages
- added it to the user settings as well

I didn't disallow names with spaces in the middle, because this is not causing issues and seems like a valid use case, e.g. for "Firstname Lastname".

cp2004 commented on src/octoprint/server/api/access.py:
```diff
@@ -163,6 +163,8 @@ def add_user():
     except users.UserAlreadyExists:
         abort(409)
+    except users.InvalidUserName:
+        abort(409)
```
HTTP 409 is not the correct status code here, because that means that there is a conflict. With username already existing, there is a conflict with the existing user, so 409.

For invalid username, I would say maybe 400 (bad request), but I'm happy if there is something better suited for that to be suggested.

```suggestion
        abort(400, "Username invalid")
```

If something other than 400 is used, then the API documentation would have to be updated as well.

040medien replied:
Resolved

cp2004 reviewed (changes requested):
Looking quite good to me. I've added a couple of comments for consistency with the other areas & API.

If you installed pre-commit in your development environment, you wouldn't have to wait for the PR checks to fix the formatting/imports etc. each time.

040medien commented on src/octoprint/plugins/corewizard/templates/corewizard_acl_wizard.jinja2:
```diff
@@ -7,10 +7,11 @@
         <label class="control-label" for="first_run_username">{{ _('Username') }}</label>
         <div class="controls">
             <input type="text" class="input-medium" data-bind="value: username, valueUpdate: 'afterkeydown', enable: !setup(), css: {disabled: setup()}">
+            <span class="help-inline" data-bind="visible: invalidUsername()">{{ _('Invalid user name') }}</span>
```
Thank you for your helpful comments @cp2004. I think I implemented them all - including renaming some functions I introduced. I did find some other mentions of "UserName" and "user name" in older code, but I didn't refactor those.

Can you please help me find out why the E2E tests are failing now? I don't see a connection to my changes.

cp2004 replied:
> Can you please help me find out why the E2E tests are failing now? I don't see a connection to my changes.

The E2E tests have been very flaky recently, just ignore them. I can try and rerun them later and maybe they'll go green.

040medien commented on src/octoprint/templates/snippets/settings/accesscontrol/users.jinja2:
```diff
@@ -79,10 +79,11 @@
                             <div class="controls">
-                                <input type="text" class="input-block-level" id="settings-usersDialogAddUserName" data-bind="disable: !$root.access.users.editor.new(), value: $root.access.users.editor.name" required>
+                                <input type="text" class="input-block-level" id="settings-usersDialogAddUserName" data-bind="disable: !$root.access.users.editor.new(), value: $root.access.users.editor.name, valueUpdate: 'afterkeydown'" required>
+                                <span class="help-inline" data-bind="visible: $root.access.users.editor.invalidUserName()">{{ _('Invalid user name') }}</span>
```
Resolved

040medien commented:
@cp2004 I can't seem to resolve the "Changes requested" above even though I addressed all your concerns. Can you please approve?

cp2004 commented:
Apologies that's because I haven't re-reviewed it - I was hoping to test it but haven't had time. It does look fine though.

I might be able to get to it but Gina will re-review when she can get to it anyway.

foosel commented:
I just pushed out 1.8.4, whether I'll *finally* get around to work through my PR review backlog depends on how that goes and what people report next. And yes, I'm indeed annoyed by all my time being taken up by 1.8.x right now 😅

foosel commented on src/octoprint/plugins/corewizard/static/js/corewizard.js:
```diff
@@ -12,12 +12,24 @@ $(function () {
 
         self.required = false;
 
+        self.invalidUsername = ko.pureComputed(function () {
+            return (
+                self.username() &&
+                self.username().trim() !== "" &&
+                self.username() !== self.username().trim()
+            );
+        });
+
```
Instead of fully duplicating the logic of `validUsername` here, this should really just be used instead, negated if needed. Especially since this is logically wrong (the opposite of `a && b && c` is `!a || !b || !c`, not `!a && !b && !c` and especially not `a && !b && !c`).

040medien replied:
The main reason I had created it in this (arguably convoluted way) was that I didn't want the wizard and the add user dialog to start with an error state. After implementing your suggested changes it shows an error until you start typing a name. Let me know if that's the expected behavior though.

<img width="549" alt="Bildschirmfoto 2022-09-28 um 6 38 44 PM" src="https://user-images.githubusercontent.com/115072/192839347-8b7456c0-ae3e-4723-
<img width="720" alt="Bildschirmfoto 2022-09-28 um 6 47 35 PM" src="https://user-images.githubusercontent.com/115072/192841720-30a004bf-c250-4ed9-a6ba-cbb5693e95d0.png">
b94f-cc0d953f6106.png">

foosel replied:
Ah... yeah, in that case I'd say go with a `validUsername` but drop the `username().trim() !== ""` requirement. Maybe we should actually just disable the Create button in that case (since it would cause an error when submitted), but not mark it as an error, that would feel more intuitive. So, `validUsername` used for error marking, `enableSubmit` (or something like that) tied to `validUsername` and `username().trim() !== ""`.

040medien replied:
I needed to solve this a little bit differently. If the username is empty, the `validUserName()` function would always returns false. So I'm allowing an empty username to be valid, but add another function `providedUsername()` to check against so the form can't be submitted.

foosel commented on src/octoprint/plugins/corewizard/static/js/corewizard.js:
```diff
@@ -26,7 +38,10 @@ $(function () {
         self.validData = ko.pureComputed(function () {
             return (
-                !self.passwordMismatch() && self.validUsername() && self.validPassword()
+                !self.invalidUsername() &&
+                !self.passwordMismatch() &&
+                self.validUsername() &&
+                self.validPassword()
```
```suggestion
                !self.passwordMismatch() &&
                self.validUsername() &&
                self.validPassword()
```

foosel commented on src/octoprint/plugins/corewizard/templates/corewizard_acl_wizard.jinja2:
```diff
@@ -7,10 +7,11 @@
 </p>{% endtrans %}
 <form class="form-horizontal" onsubmit="return false;">
-    <div class="control-group" data-bind="css: {success: validUsername()}">
+    <div class="control-group" data-bind="css: {error: invalidUsername(), success: validUsername()}">
```
```suggestion
    <div class="control-group" data-bind="css: {error: !validUsername(), success: validUsername()}">
```

foosel commented on src/octoprint/plugins/corewizard/templates/corewizard_acl_wizard.jinja2:
```diff
@@ -7,10 +7,11 @@
         <label class="control-label" for="first_run_username">{{ _('Username') }}</label>
         <div class="controls">
             <input type="text" class="input-medium" data-bind="value: username, valueUpdate: 'afterkeydown', enable: !setup(), css: {disabled: setup()}">
+            <span class="help-inline" data-bind="visible: invalidUsername()">{{ _('Invalid username') }}</span>
```
```suggestion
            <span class="help-inline" data-bind="visible: !validUsername()">{{ _('Invalid username') }}</span>
```

foosel commented on src/octoprint/static/js/app/viewmodels/access.js:
```diff
@@ -89,6 +95,7 @@ $(function () {
                     return (
                         self.editor.name() &&
                         self.editor.name().trim() &&
+                        !self.editor.invalidUsername() &&
```
Instead of double negation:

```suggestion
                        self.editor.validUsername() &&
```

foosel reviewed (changes requested):
The logic is a bit backwards, and there's too much duplication. Let's stick to defining what a valid username is and then negate that as needed. And make sure we are consistent in the definition of a valid username.

040medien commented on src/octoprint/static/js/app/viewmodels/access.js:
```diff
@@ -45,6 +45,12 @@ $(function () {
                 passwordMismatch: ko.pureComputed(function () {
                     return self.editor.password() !== self.editor.repeatedPassword();
                 }),
+                invalidUsername: ko.pureComputed(function () {
+                    return (
+                        self.editor.name() &&
+                        self.editor.name() !== self.editor.name().trim()
+                    );
+                }),
```
I think the last line would need to be
`self.editor.name() == self.editor.name().trim()`

040medien commented:
Implemented the changes. Please see my comment on the error states above.

foosel commented:
> Please see my comment on the error states above.

Good point, answered above.
